src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -31,11 +31,110 @@ public struct CallGraphNode
             public QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
         }
 
+        // TODO: 
+        // This is the method that should be invoked to verify cycles of interest,
+        // i.e. where each callable in the cycle is type parametrized.
+        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined. 
+        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        {
+            var parent = rootNode.CallableName;
+            var validResolution = TryCombineTypeResolutions(parent, out var combined, edges.Select(edge => edge.ParamResolutions).ToArray());
+            var resolvedToConcrete = combined.Values.All(res => !(res.Resolution is ResolvedTypeKind.TypeParameter tp) || tp.Item.Origin.Equals(parent));
+            return validResolution && resolvedToConcrete;
+            //var isClosedCycle = validCycle && combined.Values.Any(res => res.Resolution is ResolvedTypeKind.TypeParameter tp && EqualsParent(tp.Item.Origin));
+            // TODO: check that monomorphization correctly processes closed cycles - meaning add a test...
+        }
+
+        /// <summary>
+        /// Combines subsequent concretions as part of a nested expression, or concretions as part of a cycle in the call graph,
+        /// into a single dictionary containing the resolution for the type parameters of the specified parent callable. 
+        /// The given resolutions are expected to be ordered starting with the dictionary containing the initial mapping for the 
+        /// type parameters of the specified parent callable (the "innermost resolutions"). This mapping may potentially be to 
```

Comment:
Based on the code, it seems like it is important to have the resolutions that are guaranteed to have resolutions for native type parameters, which would be the 'outermost resolutions', come first in the ordering. Is that interpretation correct?

Reply:
No, thinking of call-like expressions, the dictionary to pass first is indeed the one of the innermost expression, and the parent would be the callable invoked in the call, like the doc comment states.